src/database-controller/src/write-merger/handler.js

```diff
@@ -0,0 +1,287 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+const createError = require('http-errors');
+const logger = require('@dbc/common/logger');
+const AsyncLock = require('async-lock');
+const DatabaseModel = require('openpaidbsdk');
+const config = require('@dbc/write-merger/config');
+const {
+  Snapshot,
+  AddOns,
+  silentSynchronizeRequest,
+  silentDeleteFramework,
+} = require('@dbc/common/framework');
+const _ = require('lodash');
+const lock = new AsyncLock({ maxPending: Number.MAX_SAFE_INTEGER });
+const databaseModel = new DatabaseModel(
+  config.dbConnectionStr,
+  config.maxDatabaseConnection,
+);
+
+/* For error handling, all handlers follow the same structure:
+
+  try{
+    ....
+  } catch (err) {
+    return next(err)
+  }
+
+  If a normal HTTP error is wanted, use `createError`.
+  e.g. return next(createError(401, 'Please login to view this page.'))
+
+  If a 500 error is wanted, throw it in-place: throw new Error('fatal error')
+
+*/
+
+async function ping(req, res, next) {
+  try {
+    res.status(200).json({ message: 'ok' });
+  } catch (err) {
+    return next(err);
+  }
+}
+
+async function postWatchEvents(req, res, next) {
+  try {
+    let snapshot = new Snapshot(req.body);
+    const frameworkName = snapshot.getName();
+    if (!frameworkName) {
+      return next(createError(400, 'Cannot find framework name.'));
+    }
+    await lock.acquire(frameworkName, async () => {
+      const oldFramework = await databaseModel.Framework.findOne({
+        attributes: ['snapshot'],
+        where: { name: frameworkName },
+      });
+      // database doesn't have the corresponding framework.
+      if (!oldFramework) {
+        if (config.retainModeEnabled) {
+          // If database doesn't have the corresponding framework,
+          // and retain mode is enabled
+          // tolerate the error and create framework in database.
```

Comment:
>tolerate the error and create framework in database. [](start = 13, length = 52)

"tolerate the error and create framework in database" ->
"retain the framework, i.e. do not delete it"

Reply:
Done.